warehouse/slave_worker.go

```diff
@@ -0,0 +1,504 @@
+
+			if ok && ((columnType != dataTypeInSchema) || (violatedConstraints.isViolated)) {
+				newColumnVal, convError := handleSchemaChange(
+					model.SchemaType(dataTypeInSchema),
+					model.SchemaType(columnType),
+					columnVal,
+				)
+
+				if convError != nil || violatedConstraints.isViolated {
+					if violatedConstraints.isViolated {
+						eventLoader.AddColumn(columnName, job.UploadSchema[tableName][columnName], violatedConstraints.violatedIdentifier)
+					} else {
+						eventLoader.AddEmptyColumn(columnName)
+					}
+
+					jr.outputFileWritersMap[discardsTable], err = jr.writer(discardsTable)
+					if err != nil {
+						return nil, err
+					}
+
+					err = jr.handleDiscardTypes(tableName, columnName, columnVal, columnData, violatedConstraints, jr.outputFileWritersMap[discardsTable])
+					if err != nil {
+						jr.logger.Errorf("[WH]: Failed to write to discards: %v", err)
+					}
+
+					jr.tableEventCountMap[discardsTable]++
+					continue
+				}
+
+				columnVal = newColumnVal
+			}
+
+			// Special handling for JSON arrays
+			// TODO: Will this work for both BQ and RS?
+			if reflect.TypeOf(columnVal) == reflect.TypeOf(interfaceSliceSample) {
+				marshalledVal, err := json.Marshal(columnVal)
+				if err != nil {
+					eventLoader.AddEmptyColumn(columnName)
+					continue
+				}
+
+				columnVal = string(marshalledVal)
+			}
+
+			eventLoader.AddColumn(columnName, job.UploadSchema[tableName][columnName], columnVal)
+		}
+
+		if err = eventLoader.Write(); err != nil {
+			return nil, err
+		}
+
+		jr.tableEventCountMap[tableName]++
+	}
+
+	jr.logger.Debugf("[WH]: Process %v bytes from downloaded staging file: %s", lineBytesCounter, job.StagingFileLocation)
+
+	jr.processingStagingFileStat.Since(processingStart)
+	jr.bytesProcessedStagingFileStat.Count(lineBytesCounter)
+
+	for _, loadFile := range jr.outputFileWritersMap {
+		if err = loadFile.Close(); err != nil {
+			jr.logger.Errorf("Error while closing load file %s : %v", loadFile.GetLoadFile().Name(), err)
+		}
+	}
+
+	uploadsResults, err := jr.uploadLoadFiles(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return uploadsResults, err
+}
+
+func (sw *slaveWorker) processClaimedAsyncJob(ctx context.Context, claimedJob pgnotifier.Claim) {
+	handleErr := func(err error, claim pgnotifier.Claim) {
+		sw.log.Errorf("[WH]: Error processing claim: %v", err)
+
+		sw.notifier.UpdateClaimedEvent(&claimedJob, &pgnotifier.ClaimResponse{
+			Err: err,
+		})
+	}
+
+	var (
+		job jobs.AsyncJobPayload
+		err error
+	)
+
+	if err := json.Unmarshal(claimedJob.Payload, &job); err != nil {
+		handleErr(err, claimedJob)
+		return
+	}
+
+	jobResult, err := sw.runAsyncJob(ctx, job)
+	if err != nil {
+		handleErr(err, claimedJob)
+		return
+	}
+
+	jobResultJSON, err := json.Marshal(jobResult)
+	if err != nil {
+		handleErr(err, claimedJob)
+		return
+	}
+
+	sw.notifier.UpdateClaimedEvent(&claimedJob, &pgnotifier.ClaimResponse{
+		Payload: jobResultJSON,
+	})
```

Comment:
[questions] what is the side effect of these calls failing? I see that we log the errors but beside that, if we don't update the claimed event, what happens next?

Reply:
If we don't update the claim event, the side effect is that the `notifier.trackUploadBatch` will wait for it to complete. Also, in case of failures, since we are just logging, there are maintenance workers who will mark them again as waiting so that it can be picked it up again by another slave.

Reply:
Got it, thanks. Does `trackUploadBatch` wait forever or we got a timeout there?

Reply:
Yeah, it waits forever for the job to be completed. Although we have 
1. maintenance workers which upon some threshold mark the job as waiting.
2. also, we have alerts on long-running tasks in cases it exceeds 2 hours.

Reply:
We can definitely improve things on pgNotifier.